src/phy/fuzz_injector.rs

```diff
@@ -0,0 +1,120 @@
+use Result;
+use phy::{self, DeviceCapabilities, Device};
+use time::Instant;
+
+// This could be fixed once associated consts are stable.
+const MTU: usize = 1536;
+
+/// Represents a fuzzer. It is expected to either generate or contain data, which it replaces bytes
+/// in the packet with.
+pub trait Fuzzer {
+    /// Modify a single packet.
+    fn fuzz_packet(&self, packet_data: &mut [u8]);
+}
+
+/// A fuzz injector device.
+///
+/// A fuzz injector is a device that alters packets traversing through it according to the
+/// directions of a guided fuzzer. It is designed to support fuzzing internal state machines inside
+/// smoltcp, and is not for production use.
+#[allow(unused)]
+#[derive(Debug)]
+pub struct FuzzInjector<D: for<'a> Device<'a>, FTx: Fuzzer, FRx: Fuzzer> {
+    inner:   D,
+    fuzz_tx: FTx,
+    fuzz_rx: FRx,
+}
+
+#[allow(unused)]
+impl<D: for<'a> Device<'a>, FTx: Fuzzer, FRx: Fuzzer> FuzzInjector<D, FTx, FRx> {
+    /// Create a fuzz injector device, which defaults to passing the packets unaltered.
+    pub fn new(inner: D, fuzz_tx: FTx, fuzz_rx: FRx) -> FuzzInjector<D, FTx, FRx> {
+        FuzzInjector { inner, fuzz_tx, fuzz_rx }
+    }
+
+    /// Return the underlying device, consuming the fuzz injector.
+    pub fn into_inner(self) -> D {
+        self.inner
+    }
+}
+
+impl<'a, D, FTx, FRx> Device<'a> for FuzzInjector<D, FTx, FRx>
+    where D: for<'b> Device<'b>,
+          FTx: Fuzzer + 'a,
+          FRx: Fuzzer + 'a
+{
+    type RxToken = RxToken<'a, <D as Device<'a>>::RxToken, FRx>;
+    type TxToken = TxToken<'a, <D as Device<'a>>::TxToken, FTx>;
+
+    fn capabilities(&self) -> DeviceCapabilities {
+        let mut caps = self.inner.capabilities();
+        if caps.max_transmission_unit > MTU {
+            caps.max_transmission_unit = MTU;
+        }
+        caps
+    }
+
+    fn receive(&'a mut self) -> Option<(Self::RxToken, Self::TxToken)> {
+        let &mut Self { ref mut inner, ref fuzz_rx, ref fuzz_tx } = self;
+        inner.receive().map(|(rx_token, tx_token)| {
+            let rx = RxToken {
+                fuzzer: fuzz_rx,
+                token:   rx_token,
+                fuzzed:  [0; MTU]
+            };
+            let tx = TxToken {
+                fuzzer: fuzz_tx,
+                token:   tx_token,
+            };
+            (rx, tx)
+        })
+    }
+
+    fn transmit(&'a mut self) -> Option<Self::TxToken> {
+        let &mut Self { ref mut inner, fuzz_rx: _, ref fuzz_tx } = self;
+        inner.transmit().map(|token| TxToken {
+            fuzzer: fuzz_tx,
+            token:   token,
+        })
+    }
+}
+
+#[doc(hidden)]
+pub struct RxToken<'a, Rx: phy::RxToken, F: Fuzzer>{
+    fuzzer: &'a F,
+    token:   Rx,
+    fuzzed: [u8; MTU]
```

Comment:
Is this buffer actually necessary? Why?

Reply:
The buffer is necessary because the buffer passed into the RxToken.consume is not mutable. I tried really hard to eliminate it, but the closure simply would not support it.

The fixed MTU was copied from the loopback example.

Reply:
Hmm, maybe we should make it mutable?

Reply:
You mean change the trait? Because the ``F`` type on the trait is ``FnOnce(&[u8])``, which is why I can't mutate the data.

That would mean a bunch of little changes scattered everywhere, not to mention any multi-threading consequences...

Reply:
> That would mean a bunch of little changes scattered everywhere, not to mention any multi-threading consequences...

Yeah. Would you mind making this change? I think there's no especially good reason RxToken doesn't already present a mutable buffer. In this context it is better thought of as "exclusive" moreso than "mutable"; any network buffer would be consumed by only one network stack anyway, and making it mutable would potentially allow doing some interesting things with it, like the equivalent of a "mangle" firewall table.

Reply:
Opened separate PR #294